src/uct/ib/rdmacm/rdmacm_cm_ep.c

```diff
@@ -149,15 +149,123 @@ ucs_status_t uct_rdmacm_cm_ep_conn_notify(uct_ep_h ep)
     return cep->status;
 }
 
+/**
+ * Allocate a reserved QPN either from the last FW object allocated,
+ * or by allocating a new one. When find a free QPN in an object, it
+ * always starts from last allocation position, to make sure the QPN
+ * always move forward to prevent stale QPN.
+ */
+static ucs_status_t
+uct_rdamcm_cm_ep_create_reserved_qpn(uct_rdmacm_cm_ep_t *cep,
+                                     struct uct_rdmacm_cm_qp_context *ctx)
+{
+    ucs_status_t status = UCS_OK;
+    struct uct_rdmacm_cm_reserved_qpn_blk *blk;
+    uint32_t qpns_per_obj;
+
+    qpns_per_obj = uct_rdmacm_cm_get_qpns_per_obj(ctx->reserved_qpn_ctx);
+
+    pthread_mutex_lock(&ctx->reserved_qpn_ctx.mutex);
+
+    blk = ucs_list_tail(&ctx->reserved_qpn_ctx.blk_list,
+                        struct uct_rdmacm_cm_reserved_qpn_blk, entry);
+    if (ucs_list_is_empty(&ctx->reserved_qpn_ctx.blk_list) ||
+        (blk->next_avail_qpn_offset >= qpns_per_obj)) {
+        status = uct_rdmacm_cm_reserved_qpn_blk_alloc(uct_rdmacm_cm_ep_get_cm(cep),
+                                                      cep->id->verbs, ctx, &blk);
+        if (status != UCS_OK) {
+            goto out;
+        }
+
+        ucs_list_add_tail(&ctx->reserved_qpn_ctx.blk_list, &blk->entry);
+    }
+
+    cep->qpn = blk->first_qpn + blk->next_avail_qpn_offset;
+    ucs_debug("created reserved qp_num: 0x%x on rdmacm_id %p",
+              cep->qpn, cep->id);
+    UCS_BITMAP_SET(blk->bmp, blk->next_avail_qpn_offset);
+    blk->next_avail_qpn_offset++;
+
+out:
+    pthread_mutex_unlock(&ctx->reserved_qpn_ctx.mutex);
+    return status;
+}
+
+/**
+ * Deallocate a reserved QPN. The FW object is destroyed only when all QPNs
+ * in this object were used and freed.
+ */
+static void uct_rdamcm_cm_ep_destroy_reserved_qpn(uct_rdmacm_cm_ep_t *cep)
+{
+    struct uct_rdmacm_cm_qp_context *ctx;
+    struct uct_rdmacm_cm_reserved_qpn_blk *blk;
+    uint32_t qpns_per_obj;
+    bool found;
+    ucs_status_t status;
+
+    if (!(cep->flags & UCT_RDMACM_CM_EP_QPN_CREATED)) {
```

Comment:
why would it be false?

Reply:
@yosefe Yes, the destroy function called in cleanup function can be called before qp created. I reported this in an email before. You suggested that we should only destroy qpn after it was created. So add this flag to indicate qpn was created or not.

Reply:
ok, pls add comment:
/* Reserved QP number is created on-demand weh nwe send the private data */

Reply:
@yosefe Added comment.